### PR TITLE
WIP: Base Breakup

### PR DESCRIPTION
**Pull request by rofinn**
**Warning**: This is just a prototype of how we **could** breakup base and is intended to help facilitate discussion on the topic.

## Summary

* Include a default REQUIRE file at `./etc/julia/REQUIRE` which includes the default packages to install.
* Provide an `@initialize(...)` macro which clones a specific version of the `Pkg.jl` repo to `./usr/share/julia/site/v0.7/Pkg.jl` (so that `using Pkg` works).
* Loads `Pkg` and runs `Pkg.init()`
* Runs `Pkg.edit()` (`@initialize(true)`) or `Pkg.resolve()` (`@initialize(false)`)

NOTE: I opted to include the code for pulling Pkg out of base, but we could choose to do this in stages (ie: introduce default packages in the REQUIRE file first and then deal with Pkg later).

## Pros

* `@initialize` allows everyone to customize their julia setup regardless of whether they're using a binary release or building from source (vs putting this logic in the `Makefile`).
* `Pkg` can be registered on METADATA, so you can do things like `Pkg.add("Pkg")` to get a newer version.
* Getting folks reading and editing a REQUIRE file (or whatever this looks like for Pkg3) could be a good introductory step.
* Uses the same tooling for default and 3rd party packages (edit the REQUIRE file and run `Pkg.resolve()`).
* These default packages could serve as a "model" for folks building their own packages.
* Could include other steps in this initialize macro (e.g., setting the `EDITOR`, generating custom system images, even setting up a juliarc).
* Reuses the complicated stuff like parsing and installing packages from the REQUIRE file and loading packages from the `./usr/share/julia/site/v0.7/` directory.

## Cons

* Pkg has a few internal dependencies in Base that don't work if Pkg isn't built into the system image (e.g., `versioninfo(true)` depends on `Pkg.status()`).
* Might not be worth it for Pkg2. I'd personally like to see a Pkg.jl repository with v2.0.0 first and then merge the Pkg3 code in for a v3.0.0 release, but the changes to Pkg3 might make a lot of the changes in Base unnecessary.

### CODE REVIEW

File: base/init.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+_pkg_require() = abspath(JULIA_HOME, "..", "etc", "julia", "REQUIRE")
+
+"""
+    initialize(interactive)
+
+Initializes a collection of core julia packages.
+If `interactive=true` you will be prompted to select the packages to include by
+editing the default REQUIRE file included in base julia.
+"""
+macro initialize(arg)
```
Comment (JeffBezanson) [inline base/init.jl]: It seems to me this can just be a function.

Comment (rofinn) [inline base/init.jl]: Yeah, I made this a macro cause I was initially trying to do `using Pkg`, but I think making it a function and keeping the `Pkg = include(...)` is the better choice.

**Comment by StefanKarpinski** [comment]
Cool, will take a look through this.

**Comment by kshyatt** [comment]
It seems this was pre-empted by the `stdlib`.

**Comment by rofinn** [comment]
Yep, thanks. Is there a plan to move `Pkg` to the `stdlib`?

**Comment by waldyrious** [comment]
> Yep, thanks. Is there a plan to move `Pkg` to the `stdlib`?

#24702

**Pull request closed without merging**